File: app/src/main/java/dev/ragnarok/fenrir/api/interfaces/IMessagesApi.java
package dev.ragnarok.fenrir.api.interfaces;

import androidx.annotation.CheckResult;

import java.util.Collection;
import java.util.List;
import java.util.Map;

import dev.ragnarok.fenrir.api.model.IAttachmentToken;
import dev.ragnarok.fenrir.api.model.Items;
import dev.ragnarok.fenrir.api.model.VKApiChat;
import dev.ragnarok.fenrir.api.model.VKApiMessage;
import dev.ragnarok.fenrir.api.model.VkApiConversation;
import dev.ragnarok.fenrir.api.model.VkApiJsonString;
import dev.ragnarok.fenrir.api.model.VkApiLongpollServer;
import dev.ragnarok.fenrir.api.model.response.AttachmentsHistoryResponse;
import dev.ragnarok.fenrir.api.model.response.ConversationDeleteResult;
import dev.ragnarok.fenrir.api.model.response.ConversationMembersResponse;
import dev.ragnarok.fenrir.api.model.response.ConversationsResponse;
import dev.ragnarok.fenrir.api.model.response.DialogsResponse;
import dev.ragnarok.fenrir.api.model.response.ItemsProfilesGroupsResponse;
import dev.ragnarok.fenrir.api.model.response.LongpollHistoryResponse;
import dev.ragnarok.fenrir.api.model.response.MessageHistoryResponse;
import dev.ragnarok.fenrir.api.model.response.MessageImportantResponse;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Single;

public interface IMessagesApi {

    @CheckResult
    Completable edit(int peerId, int messageId, String message, List<IAttachmentToken> attachments, boolean keepFwd, Boolean keepSnippets);

    @CheckResult
    Single<Boolean> removeChatMember(int chatId, int memberId);

    @CheckResult
    Single<Boolean> deleteChatPhoto(int chatId);

    @CheckResult
    Single<Boolean> addChatUser(int chatId, int userId);

    @CheckResult
    Single<List<VKApiChat>> getChat(Integer chatId, Collection<Integer> chatIds, String fields, String name_case);

    @CheckResult
    Single<ConversationMembersResponse> getConversationMembers(Integer peer_id, String fields);

    @CheckResult
    Single<Boolean> editChat(int chatId, String title);

    @CheckResult
    Single<Integer> createChat(Collection<Integer> userIds, String title);

    @CheckResult
    Single<ConversationDeleteResult> deleteDialog(int peerId);

    @CheckResult
    Single<Boolean> restore(int messageId);

    @CheckResult
    Single<Map<String, Integer>> delete(Collection<Integer> messageIds, Boolean deleteForAll, Boolean spam);

    @CheckResult
    Single<Boolean> markAsRead(Integer peerId, Integer startMessageId);

    @CheckResult
    Single<Boolean> setActivity(int peerId, boolean typing);

    @CheckResult
    Single<Items<VKApiMessage>> search(String query, Integer peerId, Long date, Integer previewLength,
                                       Integer offset, Integer count);

    @CheckResult
    Single<List<Integer>> markAsImportant(Collection<Integer> messageIds, Integer important);

    @CheckResult
    Single<LongpollHistoryResponse> getLongPollHistory(Long ts, Long pts, Integer previewLength,
                                                       Boolean onlines, String fields,
                                                       Integer eventsLimit, Integer msgsLimit,
                                                       Integer max_msg_id);

    @CheckResult
    Single<AttachmentsHistoryResponse> getHistoryAttachments(int peerId, String mediaType, String startFrom, Integer photoSizes,
                                                             Integer count, String fields);


    @CheckResult
    Single<Integer> send(Integer randomId, Integer peerId, String domain, String message,
                         Double latitude, Double longitude, Collection<IAttachmentToken> attachments,
                         Collection<Integer> forwardMessages, Integer stickerId, String payload, Integer reply_to);

    @CheckResult
    Single<DialogsResponse> getDialogs(Integer offset, Integer count, Integer startMessageId, Boolean extended, String fields);

    @CheckResult
    Single<ItemsProfilesGroupsResponse<VkApiConversation>> getConversations(List<Integer> peers, Boolean extended, String fields);

    @CheckResult
    Single<List<VKApiMessage>> getById(Collection<Integer> ids);

    @CheckResult
    Single<MessageHistoryResponse> getHistory(Integer offset, Integer count, int peerId, Integer startMessageId, Boolean rev, Boolean extended, String fields);

    @CheckResult
    Single<Items<VkApiJsonString>> getJsonHistory(Integer offset, Integer count, int peerId);

    @CheckResult
    Single<MessageImportantResponse> getImportantMessages(Integer offset, Integer count, Integer startMessageId, Boolean extended, String fields);

    @CheckResult
    Single<VkApiLongpollServer> getLongpollServer(boolean needPts, int lpVersion);

    @CheckResult
    Single<ConversationsResponse> searchConversations(String query, Integer count, Integer extended, String fields);

    @CheckResult
    Completable pin(int peerId, int messageId);

    @CheckResult
    Completable unpin(int peerId);

    @CheckResult
    Completable pinUnPinConversation(int peerId, boolean peen);

    @CheckResult
    Single<Integer> recogniseAudioMessage(Integer message_id, String audio_message_id);

    @CheckResult
    Single<Integer> setMemberRole(Integer peer_id, Integer member_id, String role);
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/api/interfaces/IMessagesApi.kt
package dev.ragnarok.fenrir.api.interfaces

import androidx.annotation.CheckResult
import dev.ragnarok.fenrir.api.model.*
import dev.ragnarok.fenrir.api.model.response.*
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Single

interface IMessagesApi {
    @CheckResult
    fun edit(
        peerId: Int,
        messageId: Int,
        message: String?,
        attachments: List<IAttachmentToken>?,
        keepFwd: Boolean,
        keepSnippets: Boolean?
    ): Completable

    @CheckResult
    fun removeChatMember(chatId: Int, memberId: Int): Single<Boolean>

    @CheckResult
    fun deleteChatPhoto(chatId: Int): Single<Boolean>

    @CheckResult
    fun addChatUser(chatId: Int, userId: Int): Single<Boolean>

    @CheckResult
    fun getChat(
        chatId: Int?,
        chatIds: Collection<Int>?,
        fields: String?,
        name_case: String?
    ): Single<List<VKApiChat>>

    @CheckResult
    fun getConversationMembers(
        peer_id: Int?,
        fields: String?
    ): Single<ConversationMembersResponse>

    @CheckResult
    fun editChat(chatId: Int, title: String?): Single<Boolean>

    @CheckResult
    fun createChat(userIds: Collection<Int?>?, title: String?): Single<Int>

    @CheckResult
    fun deleteDialog(peerId: Int): Single<ConversationDeleteResult>

    @CheckResult
    fun restore(messageId: Int): Single<Boolean>

    @CheckResult
    fun delete(
        messageIds: Collection<Int>,
        deleteForAll: Boolean?,
        spam: Boolean?
    ): Single<Map<String, Int>>

    @CheckResult
    fun markAsRead(peerId: Int?, startMessageId: Int?): Single<Boolean>

    @CheckResult
    fun setActivity(peerId: Int, typing: Boolean): Single<Boolean>

    @CheckResult
    fun search(
        query: String?, peerId: Int?, date: Long?, previewLength: Int?,
        offset: Int?, count: Int?
    ): Single<Items<VKApiMessage>>

    @CheckResult
    fun markAsImportant(messageIds: Collection<Int>, important: Int?): Single<List<Int>>

    @CheckResult
    fun getLongPollHistory(
        ts: Long?, pts: Long?, previewLength: Int?,
        onlines: Boolean?, fields: String?,
        eventsLimit: Int?, msgsLimit: Int?,
        max_msg_id: Int?
    ): Single<LongpollHistoryResponse>

    @CheckResult
    fun getHistoryAttachments(
        peerId: Int, mediaType: String?, startFrom: String?, photoSizes: Int?,
        count: Int?, fields: String?
    ): Single<AttachmentsHistoryResponse>

    @CheckResult
    fun send(
        randomId: Long?, peerId: Int?, domain: String?, message: String?,
        latitude: Double?, longitude: Double?, attachments: Collection<IAttachmentToken>?,
        forwardMessages: Collection<Int>?, stickerId: Int?, payload: String?, reply_to: Int?
    ): Single<Int>

    @CheckResult
    fun getDialogs(
        offset: Int?,
        count: Int?,
        startMessageId: Int?,
        extended: Boolean?,
        fields: String?
    ): Single<DialogsResponse>

    @CheckResult
    fun getConversations(
        peers: List<Int>,
        extended: Boolean?,
        fields: String?
    ): Single<ItemsProfilesGroupsResponse<VKApiConversation>>

    @CheckResult
    fun getById(identifiers: Collection<Int>?): Single<List<VKApiMessage>>

    @CheckResult
    fun getHistory(
        offset: Int?,
        count: Int?,
        peerId: Int,
        startMessageId: Int?,
        rev: Boolean?,
        extended: Boolean?,
        fields: String?
    ): Single<MessageHistoryResponse>

    @CheckResult
    fun getJsonHistory(offset: Int?, count: Int?, peerId: Int): Single<Items<VKApiJsonString>>

    @CheckResult
    fun getImportantMessages(
        offset: Int?,
        count: Int?,
        startMessageId: Int?,
        extended: Boolean?,
        fields: String?
    ): Single<MessageImportantResponse>

    @CheckResult
    fun getLongpollServer(needPts: Boolean, lpVersion: Int): Single<VKApiLongpollServer>

    @CheckResult
    fun searchConversations(
        query: String?,
        count: Int?,
        extended: Int?,
        fields: String?
    ): Single<ConversationsResponse>

    @CheckResult
    fun pin(peerId: Int, messageId: Int): Completable

    @CheckResult
    fun unpin(peerId: Int): Completable

    @CheckResult
    fun pinUnPinConversation(peerId: Int, peen: Boolean): Completable

    @CheckResult
    fun markAsListened(message_id: Int): Completable

    @CheckResult
    fun recogniseAudioMessage(message_id: Int?, audio_message_id: String?): Single<Int>

    @CheckResult
    fun setMemberRole(peer_id: Int?, member_id: Int?, role: String?): Single<Int>
}